### PR TITLE
Log window fixes (solution for issue 1717)

### PR DESCRIPTION
**Pull request by hunterzero99**
Used a buffer written to screen every 100ms to dramatically lower CPU use of log window.  
Limit log window to 10,000 lines to prevent RAM use from constantly increasing. 

Proposed fix for https://github.com/input-leap/input-leap/issues/1717

### CODE REVIEW

File: src/gui/src/LogWindowBase.ui [hunk 1]
```diff
@@ -6,7 +6,7 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>800</width>
+    <width>996</width>
```
Comment (sithlord48) [inline src/gui/src/LogWindowBase.ui:9]: Are these changes intended? Sadly when you edit ui files the designer loves to changes other lines you didn't want to on you.

To avoid editing this file at all in the constructor after `setupUi(this)` you can call  `m_LogOutput->setMaximumBlockSize(10000)`

File: src/gui/src/LogWindow.cpp [hunk 2]
```diff
@@ -32,33 +34,58 @@ LogWindow::LogWindow(QWidget *parent) :
...
-    if (!m_pLogOutput->toPlainText().isEmpty())
-        appendRaw("");
+    if (!buffer.isEmpty())
+        buffer += '\n';
 }
 
 void LogWindow::appendInfo(const QString& text)
 {
-    appendRaw(getTimeStamp() + " INFO: " + text);
+    buffer += getTimeStamp() + " INFO: " + text + '\n';
```
Comment (sithlord48) [inline src/gui/src/LogWindow.cpp:53]: Dont use concatantation insead do this
create a private for the widget
 `QString m_infoLine = QStringLiteral("%1 INFO: %2 \n");`
use this in place of your string 
`buffer.append(m_infoLine.arg(getTimeStamp(), text));`

using concatanation is very slow each + will cause a new check and resize of the string for the sting on the right side of the + for the string here it happens 3 times then again for the buffer... using append and/or arg subs will make this once for the string and once for the buffer.

Making our line a QStringLiteral will reduse the memory use as the template string is made at build time and stored in the binary using it at runtime costs nothing.

File: src/gui/src/LogWindow.cpp [hunk 3]
```diff
@@ -32,33 +34,58 @@ LogWindow::LogWindow(QWidget *parent) :
     // repeatedly until InputLeap is finished
     setAttribute(Qt::WA_DeleteOnClose, false);
     setupUi(this);
+    
+    // Use a timer to flush the buffer every 100 milliseconds
+    QTimer* timer = new QTimer(this);
+    connect(timer, &QTimer::timeout, this, &LogWindow::flushBuffer);
+    timer->start(100);
```
Comment (sithlord48) [inline src/gui/src/LogWindow.cpp:41]: Instad of flushing every 100msec perhaps flush when the buffer size is some size. this will prevent us from doing any extra syncs if the log is still small

File: src/gui/src/LogWindow.cpp [hunk 4]
```diff
@@ -32,33 +34,58 @@ LogWindow::LogWindow(QWidget *parent) :
...
+
+void LogWindow::flushBuffer()
+{
+    if (!buffer.isEmpty())
+    {
+        // Insert the new log content from the buffer
+        m_pLogOutput->appendPlainText(buffer);
+        buffer.clear();
+
+        // Scroll to the bottom if it was at the bottom before
```
Comment (sithlord48) [inline src/gui/src/LogWindow.cpp:79]: Pretty sure that you can set a property in the widget for this behavior

**Review by sithlord48 — CHANGES_REQUESTED** [review]
Nice work!
A few changes could make this preform even better.

File: src/gui/src/LogWindow.cpp (hunk 4 above)
Comment (prahal) [reply src/gui/src/LogWindow.cpp:79]: @hunterzero99 I considered the review and made https://github.com/prahal/input-leap/tree/wip-log-window-fixes above input-leap/master.

I disabled the scroll-to-bottom code because it only stopped scrolling to the bottom if one clicked the text. Not if one clicked the scrollbar or scrolled with the mouse.
As far as I tested the output in the log window already scrolls with new content if I do not act upon it.

I also made the buffer private.

I did not code much in cpp so I took the liberty to make the "QString m_infoLine = QStringLiteral("%1 INFO: %2 \n");" into private attributes. Maybe there is a cleaner way.

I tested this code.
The only issue I have left is that with input-leap set to DEBUG2 mode and log to file the start of the log is not saved to file.
I believe the log-to-file is made from the LogWindow content and this LogWIndow content being truncated to 10000 lines cut the start of the logs (thus the version and other critical data).

Maybe the log-to-file should be handled by your new flushBuffer instead of proxying through the LogWindow code. That would solve this blocker.

Comment (prahal) [reply src/gui/src/LogWindow.cpp:79]: @sithlord48 about setting a property to get this behavior, did you mean "scroll to the bottom if it was at the bottom before"?
Note this behavior is the default one and this code "to scroll to bottom" introduces bugs (it scrolls to bottom even if we use the mouse to move in the QPlaintTextEdit. So I removed it altogether in https://github.com/prahal/input-leap/commit/2192f7233c4196692d7a90dd24fe68a865e43e0c

@hunterzero99 do you want me to take over with a PR from https://github.com/prahal/input-leap/commit/2192f7233c4196692d7a90dd24fe68a865e43e0c bringing @sithlord48 changes to your code?

I ran input-leap with max debug2 logs and had no hangs for days. My modification looks ready to go to master to me.

Comment (hunterzero99) [reply src/gui/src/LogWindow.cpp:79]: That sounds great to me @prahal, thank you!  I haven't had a chance to build from your changes yet, but they seem good to me at a visual review.
